consumer/src/main/kotlin/streams/procedures/StreamsSinkProcedures.kt

```diff
@@ -115,64 +132,58 @@ class StreamsSinkProcedures {
         Stream.of(KeyValueResult("error", "You can use this procedure only in the LEADER or in a single instance configuration."))
     }
 
-    private fun readData(topic: String, procedureConfig: Map<String, Any>, consumerConfig: Map<String, String>): List<Any> {
+    private fun readData(topic: String, procedureConfig: Map<String, Any>, consumerConfig: Map<String, String>): Stream<StreamResult?> {
         val cfg = procedureConfig.mapValues { if (it.key != "partitions") it.value else mapOf(topic to it.value) }
         val timeout = cfg.getOrDefault("timeout", 1000).toString().toLong()
-        val consumer = createConsumer(consumerConfig, topic)
-        consumer.start()
-        val data = mutableListOf<Any>()
-        try {
-            val start = System.currentTimeMillis()
-            while ((System.currentTimeMillis() - start) < timeout) {
-                consumer.read(cfg) { _, topicData ->
-                    data.addAll(topicData.mapNotNull { it.value })
+        val data = ArrayBlockingQueue<StreamResult>(1000)
+        val tombstone = StreamResult(emptyMap<String, Any>())
+        GlobalScope.launch {
```

Comment:
this is just my lack of experience with kotlin but I don't understand most of this code block unfortunately.